rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -579,7 +579,7 @@ NodeParameters::declare_parameter(
   bool ignore_override)
 {
   std::lock_guard<std::recursive_mutex> lock(mutex_);
-  ParameterMutationRecursionGuard guard(parameter_modification_enabled_);
+  // ParameterMutationRecursionGuard guard(parameter_modification_enabled_);
```

Comment:
this recursion guard is block this PR work.

https://github.com/ros2/rclcpp/blob/26f6efa840cbac047a93dc21d055cfe9cde7270f/rclcpp/include/rclcpp/node_interfaces/node_parameters.hpp#L58-L80

i guess this is not only for this PR's blocker, but also when we create either `publisher` or `subscription` with overriding the QoS parameters enabled via parameter callback. i guess it is something user would do when the specific parameter is enabled, and then create the `subscription` with QoS override. this will be generating `rclcpp::exceptions::ParameterModifiedInCallbackException` because it eventually call `rclcpp::detail::declare_qos_parameters`.

i will submit the issue to open this discussion.